### PR TITLE
Update model viewer to support PySide6 and display units and domain.

### PR DESCRIPTION
**Pull request by eslickj**
## Changes proposed in this PR:

Make these changes in model viewer:
- Drop PyQt4 support (this probably isn't used enough to take the time to test a lot of interfaces)
- Add PySide6 (because this is now most convenient for me)
- Add units to Var, Param, and Expression
- Add domain to Var
- Display BooleanVar along with Var
- Display the BoleanVar domain as "BooleanVar"
- You can now set all values in an indexed Var or Param at once by entering parent_component value in tree. 
- Run Black

### Legal Acknowledgement

By contributing to this software project, I have read the [contribution guide](https://pyomo.readthedocs.io/en/stable/contribution_guide.html) and agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the BSD license.
2. I represent I am authorized to make the contributions and grant the license. If my employer has rights to intellectual property that includes these contributions, I represent that I have received permission to make contributions and grant the required license on behalf of that employer.

### CODE REVIEW

File: pyomo/contrib/viewer/ui.py [hunk 1]
```diff
@@ -23,46 +23,48 @@
...
     def get_ipython():
         raise AttributeError("IPython not available")
-import pyomo.contrib.viewer.report as rpt
-import pyomo.environ as pyo
 
-_log = logging.getLogger(__name__)
 
+import pyomo.contrib.viewer.report as rpt
+import pyomo.environ as pyo
 from pyomo.contrib.viewer.qt import *
```
Comment (blnicho) [inline pyomo/contrib/viewer/ui.py:41]: We've been trying to replace all uses of `import *` with explicit imports. Would you mind reworking this import statement as part of this PR?

File: pyomo/contrib/viewer/ui.py [hunk 2]
```diff
@@ -23,46 +23,48 @@
...
     class _MainWindowUI(object):
         pass
+
     class _MainWindow(object):
         pass
 
+
 if not qt_available:
-    for _err in qt_import_errors:
-        _log.error(_err)
```
Comment (blnicho) [inline pyomo/contrib/viewer/ui.py:63]: More changes from https://github.com/Pyomo/pyomo/pull/1066 that I think we want to keep.

File: pyomo/contrib/viewer/ui.py [hunk 3]
```diff
@@ -148,46 +145,51 @@ def toggle_tabs(self):
 
     def _tree_restart(self, w, cls=ModelBrowser, **kwargs):
         """
-        Start/Restart the variables window
+        Start/Restart the a tree window
```
Comment (blnicho) [inline pyomo/contrib/viewer/ui.py:148]: Typo: "the a"

File: pyomo/contrib/viewer/ui_data.py [hunk 4]
```diff
@@ -107,15 +109,25 @@ def calculate_expressions(self):
                 self.value_cache[o] = pyo.value(o, exception=False)
             except ZeroDivisionError:
                 self.value_cache[o] = "Divide_by_0"
+            try:
+                self.value_cache_units[o] = str(pyo.units.get_units(o))
+            except:
+                # If this units aren't obtainable for whatever reason, let it go.
```
Comment (blnicho) [inline pyomo/contrib/viewer/ui_data.py:115]: Typo: "If this units aren't"

File: pyomo/contrib/viewer/model_browser.py [hunk 5]
```diff
@@ -35,31 +33,35 @@
 from pyomo.contrib.viewer.qt import *
```
Comment (blnicho) [inline pyomo/contrib/viewer/model_browser.py:33]: Could you replace this `import *` with explicit imports?

File: pyomo/contrib/viewer/ui.py (hunk 1 above)
Comment (eslickj) [reply pyomo/contrib/viewer/ui.py:41]: No problem, will do.

File: pyomo/contrib/viewer/qt.py [hunk 6]
```diff
@@ -21,88 +21,123 @@
...
-qt_import_errors = []
 
 try:
-    from PyQt5 import QtCore
+    from PySide6 import QtCore
+
+    qt_available = "PySide6"
 except:
-    qt_import_errors.append("Cannot import PyQt5.QtCore")
+    _log.error("PySide6 could not be impoerted, trying PyQt5")
```
[the remark this answers is not in the document]
Comment (eslickj) [reply pyomo/contrib/viewer/qt.py:78]: Thanks, I'll fix it.  For some reason I'm having trouble getting meaningful errors while still letting the tests pass.  For example, I just get "qt not available" for a lot of things, and I end up adding stuff to the code to find out what exactly went wrong. I'll take a closer look and try to beef up the error reporting without breaking anything.

**Comment by jsiirola** [comment]
@eslickj: FYI: Jenkins jobs are failing during `pyomo download-extensions`:
```
14:47:48 Downloading pyomo extensions (attempt 1)
14:47:51 Traceback (most recent call last):
14:47:51   File "pyomo/scripting/pyomo_main.py", line 26, in <module>
14:47:51     plugin_class = entrypoint.load()
14:47:51   File "lib/python3.8/site-packages/pkg_resources/__init__.py", line 2471, in load
14:47:51     return self.resolve()
14:47:51   File "lib/python3.8/site-packages/pkg_resources/__init__.py", line 2477, in resolve
14:47:51     module = __import__(self.module_name, fromlist=['__name__'], level=0)
14:47:51   File "pyomo/contrib/viewer/pyomo_viewer.py", line 32, in <module>
14:47:51     import pyomo.contrib.viewer.qt as myqt
14:47:51   File "pyomo/contrib/viewer/qt.py", line 128, in <module>
14:47:51     from PyQt5.qt_module.QtCore import pyqtSignal as Signal
14:47:51 ModuleNotFoundError: No module named 'PyQt5.qt_module'
14:47:51 
14:47:51 During handling of the above exception, another exception occurred:
14:47:51 
14:47:51 Traceback (most recent call last):
14:47:51   File "python/bin/pyomo", line 33, in <module>
14:47:51     sys.exit(load_entry_point('Pyomo', 'console_scripts', 'pyomo')())
14:47:51   File "python/bin/pyomo", line 25, in importlib_load_entry_point
14:47:51     return next(matches).load()
14:47:51   File "lib/python3.8/importlib/metadata.py", line 77, in load
14:47:51     module = import_module(match.group('module'))
14:47:51   File "lib/python3.8/importlib/__init__.py", line 127, in import_module
14:47:51     return _bootstrap._gcd_import(name[level:], package, level)
14:47:51   File "<frozen importlib._bootstrap>", line 1014, in _gcd_import
14:47:51   File "<frozen importlib._bootstrap>", line 991, in _find_and_load
14:47:51   File "<frozen importlib._bootstrap>", line 975, in _find_and_load_unlocked
14:47:51   File "<frozen importlib._bootstrap>", line 671, in _load_unlocked
14:47:51   File "<frozen importlib._bootstrap_external>", line 843, in exec_module
14:47:51   File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
14:47:51   File "pyomo/scripting/pyomo_main.py", line 38, in <module>
14:47:51     raise ImportError(msg)
14:47:51 ImportError: Error loading pyomo.command entry point pyomo.viewer = pyomo.contrib.viewer.pyomo_viewer:
14:47:51 Original ModuleNotFoundError: No module named 'PyQt5.qt_module'
14:47:51 Traceback:
14:47:51   File "pyomo/scripting/pyomo_main.py", line 26, in <module>
14:47:51     plugin_class = entrypoint.load()
14:47:51   File "lib/python3.8/site-packages/pkg_resources/__init__.py", line 2471, in load
14:47:51     return self.resolve()
14:47:51   File "lib/python3.8/site-packages/pkg_resources/__init__.py", line 2477, in resolve
14:47:51     module = __import__(self.module_name, fromlist=['__name__'], level=0)
14:47:51   File "pyomo/contrib/viewer/pyomo_viewer.py", line 32, in <module>
14:47:51     import pyomo.contrib.viewer.qt as myqt
14:47:51   File "pyomo/contrib/viewer/qt.py", line 128, in <module>
14:47:51     from PyQt5.qt_module.QtCore import pyqtSignal as Signal
```

**Comment by eslickj** [comment]
@jsiirola, thanks!  Should be fixed.  I forgot to test with PyQt5.

**Comment by jsiirola** [comment]
Would it be possible to add a few tests to recover some of the lost code coverage?

**Comment by eslickj** [comment]
@jsiirola, sure.  So I know where to add the tests, do the tests with Qt count, or should I focus on the non-Qt tests?

**Comment by jsiirola** [comment]
@eslickj: all tests should count toward coverage.  The Jenkins jobs have Qt, so all the current `contrib/viewer` tests run there.  The GHA tests do not have Qt, so only run a subset of the tests.

File: pyomo/contrib/viewer/model_browser.py [hunk 7]
```diff
@@ -230,70 +249,121 @@ def _get_value_callback(self):
...
+            try:
+                for o in self.data.values():
+                    o.setub(val)
+            except:
+                return
 
     def _set_active_callback(self, val):
-        if not val or val == "False" or val == "false" or val == "0" or \
-            val == "f" or val == "F":
+        if (
```
Comment (mrmundt) [inline pyomo/contrib/viewer/model_browser.py:359]: Suggestion: change this to:
`not val or val in [ ... list of terms ...]`

File: pyomo/contrib/viewer/model_browser.py [hunk 8]
```diff
@@ -310,8 +380,14 @@ def _set_active_callback(self, val):
             return
 
     def _set_fixed_callback(self, val):
-        if not val or val == "False" or val == "false" or val == "0" or \
-            val == "f" or val == "F":
+        if (
```
Comment (mrmundt) [inline pyomo/contrib/viewer/model_browser.py:383]: Suggestion: change this to:
`not val or val in [ ... list of terms ...]`

**Review by mrmundt — CHANGES_REQUESTED** [review]
Minor change requested.

**Review by mrmundt — APPROVED** [review]

**Comment by mrmundt** [comment]
@eslickj - I believe the Jenkins failure was erroneous. I'm re-running those jobs to make sure.

**Comment by eslickj** [comment]
I see I'm still a bit short on testing.  I think I have it up high enough, but I also reworked the application that uses the Jupyter qtconsole.  Hopfully I didn't mess anything up and this hits the testing targets.

**Review by jsiirola — APPROVED** [review]

**Pull request merged**